lib/Interpreter/CIFactory.cpp

```diff
@@ -338,9 +339,367 @@ static bool getVisualStudioDir(std::string &path) {
+    GET_ISYSROOT_VER(10, 5);
+  #elif defined(MAC_OS_X_VERSION_10_4)
+    GET_ISYSROOT_VER(10, 4);
+  #elif defined(MAC_OS_X_VERSION_10_3)
+    GET_ISYSROOT_VER(10, 3);
+  #elif defined(MAC_OS_X_VERSION_10_2)
+    GET_ISYSROOT_VER(10, 2);
+  #elif defined(MAC_OS_X_VERSION_10_1)
+    GET_ISYSROOT_VER(10, 1);
+  #else // MAC_OS_X_VERSION_10_0
+    GET_ISYSROOT_VER(10, 0);
+  #endif
+
+#undef GET_ISYSROOT_VER
+
+  // Scan the SDKs directory and use the latest
+  std::error_code ec;
+  std::vector<std::string> srtd;
+  for (fs::directory_iterator it(SDKs, ec), e; it != e; it.increment(ec))
+    srtd.push_back(it->path());
+  if (!srtd.empty()) {
+    std::sort(srtd.begin(), srtd.end(), std::greater<std::string>());
+    sysRoot.swap(srtd[0]);
+    return true;
+  }
+
+  return false;
+}
+
+#endif // __APPLE__, _MSC_VER
+
+using namespace clang;
+using namespace cling;
 
 namespace {
+  // This function isn't referenced outside its translation unit, but it
+  // can't use the "static" keyword because its address is used for
+  // GetMainExecutable (since some platforms don't support taking the
+  // address of main, and some platforms can't implement GetMainExecutable
+  // without being given the address of a function in the main executable).
+  std::string GetExecutablePath(const char *Argv0) {
+    // This just needs to be some symbol in the binary; C++ doesn't
+    // allow taking the address of ::main however.
+    void *MainAddr = (void*) intptr_t(GetExecutablePath);
+    return llvm::sys::fs::getMainExecutable(Argv0, MainAddr);
+  }
+
+  class AdditionalArguments {
+    typedef std::vector< std::pair<const char*,std::string> > container_t;
+    container_t m_Saved;
+
+  public:
+    
+    void addArgument(const char* arg, std::string value = std::string()) {
+      m_Saved.push_back(std::make_pair(arg,std::move(value)));
+    }
+    container_t::const_iterator begin() const { return m_Saved.begin(); }
+    container_t::const_iterator end() const { return m_Saved.end(); }
+    bool empty() const { return m_Saved.empty(); }
+  };
+
+#ifndef _MSC_VER
+
+  static void CompilerIncludePaths(const char* Compiler,
+                                   llvm::SmallVectorImpl<char>& Buf,
+                                   AdditionalArguments& Args) {
+    std::string CppInclQuery("LC_ALL=C ");
+    CppInclQuery.append(Compiler);
+    CppInclQuery.append(" -xc++ -E -v /dev/null 2>&1 >/dev/null "
+                        "| awk '/^#include </,/^End of search"
+                        "/{if (!/^#include </ && !/^End of search/){ print }}' "
+                        "| GREP_OPTIONS= grep -E \"(c|g)\\+\\+\"");
+
+    if (FILE *pf = ::popen(CppInclQuery.c_str(), "r")) {
+      Buf.resize(Buf.capacity_in_bytes());
+      while (fgets(&Buf[0], Buf.capacity_in_bytes(), pf) && Buf[0]) {
+        size_t lenbuf = strlen(&Buf[0]);
+        Buf[lenbuf - 1] = 0; // remove trailing \n
+        // Skip leading whitespace:
+        const char *start = &Buf[0];
```

Comment:
On 16/08/16 13:01, marsupial wrote:
>
> In lib/Interpreter/CIFactory.cpp 
> <https://github.com/root-mirror/cling/pull/26#discussion_r74914489>:
>
> > +                                   llvm::SmallVectorImpl<char>& Buf,
> > +                                   AdditionalArguments& Args) {
> > +    std::string CppInclQuery("LC_ALL=C ");
> > +    CppInclQuery.append(Compiler);
> > +    CppInclQuery.append(" -xc++ -E -v /dev/null 2>&1 >/dev/null "
> > +                        "| awk '/^#include </,/^End of search"
> > +                        "/{if (!/^#include </ && !/^End of search/){ print }}' "
> > +                        "| GREP_OPTIONS= grep -E \"(c|g)\\+\\+\"");
> > +
> > +    if (FILE *pf = ::popen(CppInclQuery.c_str(), "r")) {
> > +      Buf.resize(Buf.capacity_in_bytes());
> > +      while (fgets(&Buf[0], Buf.capacity_in_bytes(), pf) && Buf[0]) {
> > +        size_t lenbuf = strlen(&Buf[0]);
> > +        Buf[lenbuf - 1] = 0; // remove trailing \n
> > +        // Skip leading whitespace:
> > +        const char *start = &Buf[0];
>
> SmallString::SmallString copies.
> SmallString::str() returns a StringRef to the copy.
> http://llvm.org/docs/doxygen/html/SmallString_8h_source.html#00035
>

Not sure I understand. I am suggesting to replace Buf's type from 
SmallVectorImpl to SmallString.
>
> —
> You are receiving this because you were assigned.
> Reply to this email directly, view it on GitHub 
> <https://github.com/root-mirror/cling/pull/26/files/fb0e73d441fb1f128482dc90496ad43c1d38711c#r74914489>, 
> or mute the thread 
> <https://github.com/notifications/unsubscribe-auth/AGNuUx13_jaumZbVhRz9iWTVnAr3C-tGks5qgZiFgaJpZM4JS_ds>.
>